### PR TITLE
fix: preserve url fragment when redirectBack is used

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change

Preserve url fragment when redirectBack is used.

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.
-   [ ] If I added a new recipe, I also added the recipe entry point into the `rollup.config.mjs`

### CODE REVIEW

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/utils.ts [hunk 1]
```diff
@@ -72,26 +72,35 @@ export function getURLHash(): string {
...
-        // Prevent Open redirects by normalising path.
         try {
-            const normalisedURLPath = new NormalisedURLPath(param).getAsStringDangerous();
-            const pathQueryParams = param.split("?")[1] !== undefined ? `?${param.split("?")[1]}` : "";
-            const pathWithQueryParams = normalisedURLPath + pathQueryParams;
+            let url: URL;
+            try {
+                url = new URL(redirectToPath);
+            } catch (error) {
+                const fakeDomain = redirectToPath.startsWith("/") ? "http:localhost" : "http://localhost/";
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/utils.ts:84]: This did work as excepted but I have fixed it in both the places.
image

File: lib/ts/utils.ts [hunk 2]
```diff
@@ -72,26 +72,35 @@ export function getURLHash(): string {
...
+                url = new URL(`${fakeDomain}${redirectToPath}`);
+            }
 
-            // Ensure a leading "/" if `normalisedUrlPath` is empty but `pathWithQueryParams` is not to ensure proper redirection.
+            // Prevent Open redirects by normalising path.
+            const normalisedURLPath = new NormalisedURLPath(redirectToPath).getAsStringDangerous();
+            const pathQueryParams = url.search || "";
+            const pathHash = url.hash || "";
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/utils.ts:91]: done

File: test/end-to-end/emailverification.test.js [hunk 3]
```diff
@@ -281,14 +285,16 @@ describe("SuperTokens Email Verification", function () {
...
-            assert.deepStrictEqual(urlWithQP, "/redirect-here?foo=bar");
+            // check that we are in /redirect-here?foo=bar#cell=4,1-6,2
+            const urlWithQP = await page.evaluate(
+                () => window.location.pathname + window.location.search + window.location.hash
+            );
+            assert.deepStrictEqual(urlWithQP, "/redirect-here?foo=bar#cell=4,1-6,2");
         });
 
-        it("Should redirect to verify email screen on successful sign up when mode is REQUIRED and email is not verified and then post verification should redirect with original redirectPath (w/o leading slash) and newUser", async function () {
+        it("Should redirect to verify email screen on successful sign up when mode is REQUIRED and email is not verified and then post verification should redirect with original redirectPath (only fragment) and newUser", async function () {
```
[the remark this answers is not in the document]
Comment (anku255) [reply test/end-to-end/emailverification.test.js:348]: Alright. I have added two new tests in each file. One for query params + fragment and one just fragment.

File: test/end-to-end/passwordless.test.js [hunk 4]
```diff
@@ -781,6 +785,30 @@ export function getPasswordlessTestCases({ authRecipe, logId, generalErrorRecipe
...
+                const device = await getPasswordlessDevice(loginAttemptInfo);
+                await setInputValues(page, [{ name: "userInputCode", value: device.codes[0].userInputCode }]);
+                await submitForm(page);
+
+                await page.waitForNavigation({ waitUntil: "networkidle0" });
+
+                const { pathname, search, hash } = await page.evaluate(() => window.location);
+                assert.deepStrictEqual(pathname + search + hash, "/#cell=4,1-6,2");
+            });
+
```
[the remark this answers is not in the document]
Comment (anku255) [reply test/end-to-end/passwordless.test.js]: done

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
 
+## [0.37.1] - 2024-02-21
```
Comment (porcellus) [inline CHANGELOG.md:8]: This needs to be updated (plus there is a merge conflict)

File: test/end-to-end/passwordless.test.js [hunk 6]
```diff
@@ -1317,6 +1369,29 @@ export function getPasswordlessTestCases({ authRecipe, logId, generalErrorRecipe
                 ]);
             });
 
+            it("Successful signin w/ stored redirectToPath (only fragment) and email verification", async function () {
```
Comment (porcellus) [inline test/end-to-end/passwordless.test.js:1372]: Isn't this a duplicate? What am I missing?

File: test/end-to-end/signup.test.js [hunk 7]
```diff
@@ -170,6 +170,52 @@ describe("SuperTokens SignUp", function () {
...
+            await page.waitForNavigation({ waitUntil: "networkidle0" });
+            let text = await getAuthPageHeaderText(page);
+            let { pathname: pathAfterRedirectToAuth, href: hrefAfterRedirectToAuth } = await page.evaluate(
+                () => window.location
+            );
+
+            const url = new URL(hrefAfterRedirectToAuth);
+            const redirectToPath = url.searchParams.get("redirectToPath");
+
+            assert.equal(pathAfterRedirectToAuth, "/auth/");
```
Comment (porcellus) [inline test/end-to-end/signup.test.js:213]: please use strictEqual to compare strings.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: test/end-to-end/passwordless.test.js (hunk 6 above)
Comment (anku255) [reply test/end-to-end/passwordless.test.js:1372]: This for "links" and the other one is for code.

File: CHANGELOG.md (hunk 5 above)
Comment (anku255) [reply CHANGELOG.md]: fixed

**Review by porcellus — APPROVED** [review]

**Pull request merged**